### PR TITLE
Add test and docs around using BASIC auth with KSQL.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 
Added a integration test to cover configuring the client and server for authenticating users using HTTP BASIC authentication, i.e. username and password.

Added documentation to the same effect.  The documentation is very Java centric at the moment. But its a starting point.

### Testing done 
Added test + manual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/installation/server-config/security.rst [hunk 1]
```diff
@@ -3,8 +3,11 @@
 Configuring Security for KSQL
 =============================
 
-KSQL supports many of the security features of both Apache Kafka and the |sr|.
+KSQL supports authentication on its http endpoints and also supports many of the security features
+of the other services it communicates withof both Apache Kafka and the |sr|.
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: typo: "withof"

File: docs/installation/server-config/security.rst [hunk 2]
```diff
@@ -3,8 +3,11 @@
...
 
+- KSQL supports BASIC http authentication on its RESTful and Websocket HTTP endpoints, i.e. the
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: Replace "i.e." with something like "... endpoints, which means that the..." (Better for localization, if we ever decide to do that.)

File: docs/installation/server-config/security.rst [hunk 3]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
 .. contents:: Table of Contents
     :local:
 
+Configuring KSQL for BASIC HTTP Authentication
+----------------------------------------------
+KSQL can be configured to require users to authenticate using a username and password via the BASIC
+HTTP authentication mechanism.
+
+NOTE: If using BASIC authentication it is highly recommended that you configure the KSQL server to
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: Use `.. note:: ` syntax.

File: docs/installation/server-config/security.rst [hunk 4]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+use SSL to secure communication, as the BASIC protocol passes credentials in plain text.
+
+Use the following to configure the KSQL server to require authentication:
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: "Use the following settings..."

File: docs/installation/server-config/security.rst [hunk 5]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+    authentication.method=BASIC
+    authentication.roles=some-ksql-cluster-id
+    authentication.realm=KsqlServer-Props
+    java.security.auth.login.config=/path/to/the/jaas_config.file
+
+The ``authentication.roles`` config defines a comma separated list of user roles. To be authorized
+to use the KSQL server an authenticated user must belong to at least one of these roles.
+
+The ``authentication.realm`` config must match a section with in ``jaas_config.file``, which will
+define how the server will authenticate users:
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: "which defines how the server authenticates users..."

File: docs/installation/server-config/security.rst [hunk 6]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+    KsqlServer-Props {
+      org.eclipse.jetty.jaas.spi.PropertyFileLoginModule required
+      file="/path/to/password-file"
+      debug="false";
+    };
+
+The example ``jaas_config.file`` above uses the Jetty ``PropertyFileLoginModule``, which itself
+authenticates users by checking for their credentials in a password file.
+
+You can also use other implementations of the standard Java ``LoginModule`` interface, e.g.
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: Replace "e.g." with "like" -- I know, I know, but casual style is all the rage these days.

File: docs/installation/server-config/security.rst [hunk 7]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+
+.. code:: bash
+
+    fred: OBF:1w8t1tvf1w261w8v1w1c1tvn1w8x,user,admin
+    harry: changeme,user,developer
+    tom: MD5:164c88b302622e17050af52c89945d44,user
+    dick: CRYPT:adpexzg3FUZAk,admin,ksq-user
+
+The password hash for a user can be obtained by using the ``org.eclipse.jetty.util.security.Password``
+utility, e.g.
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: Replace "e.g.", with "for example" or just a colon.

File: docs/installation/server-config/security.rst [hunk 8]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+
+.. code:: bash
+
+    > bin/ksql-run-class org.eclipse.jetty.util.security.Password fred letmein
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: Consider breaking this up into a block for the command-line and a separate "Your output should resemble:" block.

File: docs/installation/server-config/security.rst [hunk 9]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+    MD5:0d107d09f5bbe40cade3de5c71e9e9b7
+    CRYPT:frd5btY/mvXo6
+
+The output from the command is the password encrypted using different mechanisms, starting with
+plain text.
+
+-------------------
+Configuring the Cli
+-------------------
+If the KSQL server is configured to use BASIC authentication Cli instances will need to be
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: Add a comma: "authentication, Cli ..."

File: docs/installation/server-config/security.rst [hunk 10]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+configured with suitable valid credentials.  Credentials can be passed when starting the Cli using
+the ``--user`` and ``--password`` commandline arguments e.g.
```
Comment (JimGalasyn) [inline docs/installation/server-config/security.rst]: "command-line arguments, for example:"

File: docs/installation/server-config/security.rst [hunk 11]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+authenticates users by checking for their credentials in a password file.
+
+You can also use other implementations of the standard Java ``LoginModule`` interface, e.g.
+``JDBCLoginModule`` for reading credentials from a database or the ``LdapLoginModule``.
+
+The file parameter is the location of the properties file, The format is:
+
+.. code:: bash
+
+    <username>: <password-hash>[,<rolename> ...]
```
Comment (rodesai) [inline docs/installation/server-config/security.rst]: Out of curiosity - what's the purpose of having multiple roles in the current KSQL? Does the access get logged somewhere w/ the role so its at least useful for auditing?

File: ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+  private boolean canMakeCliRequest(final String username, final String password) {
+    try (KsqlRestClient restClient = new KsqlRestClient(findHttpListener().toString())) {
+
+      if (!username.isEmpty()) {
+        restClient.setupAuthenticationCredentials(username, password);
+      }
+
+      final RestResponse<ServerInfo> response = restClient.getServerInfo();
+      return response.isSuccessful();
```
Comment (rodesai) [inline ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java]: This should be more selective - what if we get an unexpected response status?

File: ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    assertThat(canMakeCliRequest(USER_WITHOUT_ACCESS, USER_WITHOUT_ACCESS_PWD), is(false));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseWsWithValidCredsIfUserHasNoAccessToThisCluster() throws Exception {
+    assertThat(makeWsRequest(USER_WITHOUT_ACCESS, USER_WITHOUT_ACCESS_PWD), is(Code.UNAUTHORIZED));
+  }
+
+  private boolean canMakeCliRequest(final String username, final String password) {
+    try (KsqlRestClient restClient = new KsqlRestClient(findHttpListener().toString())) {
```
Comment (rodesai) [inline ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java]: final

File: ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java (hunk 12 above)
Comment (big-andy-coates) [reply ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java]: What do you mean? The restClient doesn't expose anything else.  `isSuccessful` means the status code was a success code and we got the `ServerInfo` data. From the point of view of this auth specific test, I think that is sufficient.

File: docs/installation/server-config/security.rst [hunk 14]
```diff
@@ -3,8 +3,11 @@
 Configuring Security for KSQL
 =============================
 
-KSQL supports many of the security features of both Apache Kafka and the |sr|.
+KSQL supports authentication on its http endpoints and also supports many of the security features
```
Comment (miguno) [inline docs/installation/server-config/security.rst]: Nit: http -> HTTP

File: docs/installation/server-config/security.rst (hunk 2 above)
Comment (miguno) [inline docs/installation/server-config/security.rst]: Nit: "BASIC http authentication" -> "Basic HTTP authentication"

File: docs/installation/server-config/security.rst [hunk 15]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
 .. contents:: Table of Contents
     :local:
 
+Configuring KSQL for BASIC HTTP Authentication
```
Comment (miguno) [inline docs/installation/server-config/security.rst]: Nit: "BASIC HTTP Authentication" -> "Basic HTTP Authentication"

File: docs/installation/server-config/security.rst (hunk 3 above)
Comment (miguno) [reply docs/installation/server-config/security.rst]: Not: BASIC authentication -> basic authentication

File: docs/installation/server-config/security.rst [hunk 16]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+----------------------------------------------
+KSQL can be configured to require users to authenticate using a username and password via the BASIC
```
Comment (miguno) [inline docs/installation/server-config/security.rst]: BASIC HTTP authentication -> Basic HTTP authentication

File: docs/installation/server-config/security.rst [hunk 17]
```diff
@@ -22,6 +25,83 @@ file and then :ref:`start the KSQL server <start_ksql-server>` with your configu
...
+HTTP authentication mechanism.
+
+NOTE: If using BASIC authentication it is highly recommended that you configure the KSQL server to
+use SSL to secure communication, as the BASIC protocol passes credentials in plain text.
```
Comment (miguno) [inline docs/installation/server-config/security.rst]: More BASIC! :-)

File: docs/installation/server-config/security.rst (hunk 9 above)
Comment (miguno) [inline docs/installation/server-config/security.rst]: BASIC authentication -> Basic HTTP authentication

File: ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java (hunk 13 above)
Comment (big-andy-coates) [reply ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java]: This is `final` by definition. The use of `final` here would be superfluous.

File: docs/installation/server-config/security.rst (hunk 1 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done.

File: docs/installation/server-config/security.rst (hunk 3 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done.

File: docs/installation/server-config/security.rst (hunk 4 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done

File: docs/installation/server-config/security.rst (hunk 5 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done

File: docs/installation/server-config/security.rst (hunk 6 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done.

File: docs/installation/server-config/security.rst (hunk 7 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done.

File: docs/installation/server-config/security.rst (hunk 8 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done.

File: docs/installation/server-config/security.rst (hunk 9 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done

File: docs/installation/server-config/security.rst (hunk 10 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done

File: docs/installation/server-config/security.rst [hunk 18]
```diff
@@ -39,7 +119,7 @@ For example, a trustStore is required if the Schema Registry's SSL certificates
 the JVM by default; a keyStore is required if the Schema Registry requires mutual authentication.
 
 SSL configuration for communication with the Schema Registry can be supplied using none-prefixed,
-e.g. `ssl.truststore.location`, or prefixed e.g. `ksql.schema.registry.ssl.truststore.location`,
+e.g. ``ssl.truststore.location``, or prefixed e.g. ``ksql.schema.registry.ssl.truststore.location``,
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: done.

File: docs/installation/server-config/security.rst (hunk 14 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: Done (everywhere)

File: docs/installation/server-config/security.rst (hunk 2 above)
Comment (big-andy-coates) [reply docs/installation/server-config/security.rst]: Done (everywhere)

File: ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java (hunk 12 above)
Comment (rodesai) [reply ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java]: The KsqlErrorMessage you get back should have an error code you can use to distinguish between auth errors and other errors.

Comment (big-andy-coates) [reply ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java]: Ah, with you - I'll sort that now.

**Pull request merged**